### ISSUE TITLE
Add a `dtype` option for `load_from_checkpoint`

### ISSUE THREAD

**Issue by arijit-hub** [Lightning-AI/pytorch-lightning#20833]
### Description & Motivation

Hi,
It would be nice to have a `dtype` argument for `load_from_checkpoint` along with the very cool `map_location`. It would allow the user to set the `dtype` automatically via `load_from_checkpoint` without the manual need of doing `.to(dtype)`.

### Pitch

The fix is should be pretty straightforward. Currently we have:
```python
def load_from_checkpoint(
        cls,
        checkpoint_path: Union[_PATH, IO],
        map_location: _MAP_LOCATION_TYPE = None,
        hparams_file: Optional[_PATH] = None,
        strict: Optional[bool] = None,
        **kwargs: Any,
    ) -> Self:
...
```

It would just get one more argument:
```python
def load_from_checkpoint(
        cls,
        checkpoint_path: Union[_PATH, IO],
        map_location: _MAP_LOCATION_TYPE = None,
        type:torch.dtype=None,
        hparams_file: Optional[_PATH] = None,
        strict: Optional[bool] = None,
        **kwargs: Any,
    ) -> Self:
...
```

Given this actually points to `_load_from_checkpoint` https://github.com/Lightning-AI/pytorch-lightning/blob/3d398240d2f62f2ad05e9eff557d2d5cb44f235c/src/lightning/pytorch/core/saving.py#L53

We can add our extra `dtype` argument here too and can easily change this line https://github.com/Lightning-AI/pytorch-lightning/blob/3d398240d2f62f2ad05e9eff557d2d5cb44f235c/src/lightning/pytorch/core/saving.py#L99

to `return model.to(dtype).to(device)`

Hope this is taken into consideration :)

**Comment by matsumotosan** [comment]
@arijit-hub I could be misunderstanding the motivation for this, but if it's a simple mismatch between the model parameters and data, would a conversion of the data's type suffice?

If you'd like to convert the model type in the context of post-training quantization, there is an [example](https://lightning.ai/docs/pytorch/stable/advanced/post_training_quantization.html) that might be helpful.

**Comment by arijit-hub** [comment]
Hi @matsumotosan ,

Thanks for the reference :)

For me, the motivation comes from generative models, where you might want to sample faster using float16 for a model trained in bfloat16 or float32. 

I personally felt its much more simpler to do just `.to()` to typecast into different datatypes. Do you feel that the post-quantization examples that you shared are more practical? If that's the case, I will close the issue :D 

Best, 
Arijit

**Comment by matsumotosan** [comment]
@arijit-hub Just to preface, I'm not well-informed on quantization for generative models, so please take my word with a grain of salt.

It is definitely simpler to use `.to()` to cast the model parameters, but the loss in precision can affect the model's performance. If you have a specific metric that needs to be maintained, then using something like post-training quantization can help you mitigate some of that decrease in performance while getting the boost in inference speed and memory efficiency.

**Comment by deependujha** [comment]
Hi @arijit-hub @matsumotosan @gaurav0107,
 
Small terminology note first: casting weights to fp16/bf16 here is a precision/dtype conversion, not quantization, quantization maps floats to lower-bit integer representations with scale/zero-point and usually calibration, which is a different code path.

On the feature itself, I'd lean toward dropping it. For the training path it's redundant, since `Trainer(precision=...)` already owns dtype/autocast semantics. The only remaining case is the no-Trainer inference/sampling path, but even there `dtype=` instead of `.to(dtype)` is the same one-liner with no real UX improvement. Against that small convenience, it adds a third place to specify dtype with different (eager, full-cast) semantics from the `Trainer` precision plugins, which adds redundant API for no benefit.

My suggestion would be to close this rather than merge.
